### PR TITLE
Document new asset system

### PR DESCRIPTION
**Pull request by frantic**
cc @mkonicek @martinbigio 
@brentvatne mind doing a quick proofreading? :)

image

### CODE REVIEW

File: docs/Images.md [hunk 1]
```diff
@@ -1,54 +1,79 @@
...
-When your entire codebase respects this convention, you're able to do interesting things like automatically packaging the assets that are being used in your app. Note that in the current form, nothing is enforced, but it will be in the future.
+And `button.js` code contains
 
-### Adding Static Resources to your iOS App using Images.xcassets
+```javascript
+<Image source={require('./img/check.png')} />
+```
 
-![](/react-native/img/StaticImageAssets.png)
+Packager will bundle and server image corresponding to device's screen density, e.g. on iPhone 5s `check@2x.png` will be used, on Nexus 5 – `check@3x.png`. If there is no image matching screen density, closest best option will be selected.
```
Comment (brentvatne) [inline docs/Images.md]: > bundle and server image

bundle and serve the image

> If there is no image matching screen density, closest

If there is no image matching the screen density, the closest

File: docs/Images.md [hunk 2]
```diff
@@ -1,54 +1,79 @@
...
->
-> Any time you add a new resource to `Images.xcassets` you will need to re-build your app through Xcode before you can use it - a reload from within the simulator is not enough.
+Here are some benefits that you get:
 
-*This process is currently being improved, a much better workflow will be available shortly.*
+1. Same system on iOS and Android
+2. Images live in the same folder as your JS code. Components are self-contained.
+3. No global namespace, i.e. you don't have worry about name collisions.
+4. Only the images that are actually used will be packaged into your app.
+5. Adding and changing images doesn't require app recompilation, just refresh simulator as you normally do.
```
Comment (brentvatne) [inline docs/Images.md]: > just refresh simulator

just refresh the simulator

File: docs/Images.md [hunk 3]
```diff
@@ -1,54 +1,79 @@
...
+6. Packager knows image size, no need to duplicate it in the code.
```
Comment (brentvatne) [inline docs/Images.md]: > Packager knows image size

The packager knows the image dimensions

File: docs/Images.md [hunk 4]
```diff
@@ -1,54 +1,79 @@
...
+7. Images can be distributed via [npm](https://www.npmjs.com/) packages.
 
-> **NOTE**: PNG images are required when loading with `require('image!my-icon')`
->
-> At this time, only PNG images are supported in iOS. There is an [issue](https://github.com/facebook/react-native/issues/646) that is currently addressing this bug. In the meantime a quick fix is to rename your files to my-icon.png or to use the `uri` property like: `source={{ uri: 'my-icon' }}` instead of `require()`.
+Note that in order for this to work, image name in `require` has to be known statically.
```
Comment (brentvatne) [inline docs/Images.md]: > image name in `require`

the image name in `require`

File: docs/Images.md [hunk 5]
```diff
@@ -90,7 +115,7 @@ In React Native, one interesting decision is that the `src` attribute is named `
 <Image source={{uri: 'something.jpg'}} />
 ```
 
-On the infrastructure side, the reason is that it allows us to attach metadata to this object. For example if you are using `require('image!icon')`, then we add an `isStatic` attribute to flag it as a local file (don't rely on this fact, it's likely to change in the future!). This is also future proofing, for example we may want to support sprites at some point, instead of outputting `{uri: ...}`, we can output `{uri: ..., crop: {left: 10, top: 50, width: 20, height: 40}}` and transparently support spriting on all the existing call sites.
+On the infrastructure side, the reason is that it allows us to attach metadata to this object. For example if you are using `require('./my-icon.png')`, then we add information about it's actual location and size (don't rely on this fact, it might change in the future!). This is also future proofing, for example we may want to support sprites at some point, instead of outputting `{uri: ...}`, we can output `{uri: ..., crop: {left: 10, top: 50, width: 20, height: 40}}` and transparently support spriting on all the existing call sites.
```
Comment (brentvatne) [inline docs/Images.md]: > it's actual location

its actual location

File: docs/Images.md [hunk 6]
```diff
@@ -1,54 +1,79 @@
...
 ```javascript
-// GOOD
-<Image source={require('image!my-icon')} />
+<Image source={require('./my-icon.png')} />
+```
 
-// BAD
-var icon = this.props.active ? 'my-icon-active' : 'my-icon-inactive';
-<Image source={require('image!' + icon)} />
+Image name is resolved the same way JS modules are resolved. In the example above the packager will look for `my-icon.png` in the same folder as the component that requires it. Also if you have `my-icon.ios.png` and `my-icon.android.png`, the packager will pick the file depending on the platform you are running on.
```
Comment (brentvatne) [inline docs/Images.md]: > Image name is

The image name is

**Comment by brentvatne** [comment]
Looks amazing @frantic! After those small grammar fixes let's :shipit: :rocket:

**Comment by chirag04** [comment]
:+1: This means no need to add images to xcode now?

**Comment by chirag04** [comment]
@frantic  would it consider 1.5x and 4x for android?

**Comment by frantic** [comment]
@chirag04 yes, see https://github.com/facebook/react-native/blob/master/private-cli/src/bundle/getAssetDestPathAndroid.js#L13, we map `@Nx` to `whatever-dpi` on Android.

File: docs/Images.md [hunk 7]
```diff
@@ -1,54 +1,79 @@
...
-![](/react-native/img/StaticImageAssets.png)
+Packager will bundle and server image corresponding to device's screen density, e.g. on iPhone 5s `check@2x.png` will be used, on Nexus 5 – `check@3x.png`. If there is no image matching screen density, closest best option will be selected.
 
-> **NOTE**: App build required for new resources
->
-> Any time you add a new resource to `Images.xcassets` you will need to re-build your app through Xcode before you can use it - a reload from within the simulator is not enough.
+Here are some benefits that you get:
 
-*This process is currently being improved, a much better workflow will be available shortly.*
+1. Same system on iOS and Android
```
Comment (martinbigio) [inline docs/Images.md]: nit: missing `.` at the end of the sentence

**Comment by frantic** [comment]
Thanks for the review @brentvatne, I'm really bad with `the`.
Let's merge this when we cut the next release!

**Comment by ptomasroos** [comment]
Oh my god this is good news! High five everyone involved!

**Comment by mkonicek** [comment]
Looks awesome Alex!

**Comment by satya164** [comment]
That's super awesome!

**Comment by auser** [comment]
VERY awesome!

**Comment by negativetwelve** [comment]
This is great! Just curious, are there docs already written / upcoming that document how to release apps using the new asset system? I noticed the `--assets-dest` flag got added to the packager but didn't see how that was supposed to be used relative to the compiled bundle.

**Comment by martinbigio** [comment]
I'll submit a pr tomorrow or on Thursday to update the docs for the cli :)

**Comment by negativetwelve** [comment]
awesome, thanks @martinbigio!

**Comment by shishircc** [comment]
If I use react-native bundle command line to bundle the js code, the iPhone app run through Xcode is unable to show image.  
simulator screen shot 28 oct 2015 12 06 08 pm

**Comment by frantic** [comment]
Hey @shishircc, note that for this system to work we need to add a step to Xcode build process that would collect all required images and package them into the app. See https://github.com/facebook/react-native/commit/be70e32de2c417a467af902e7f4e2a0058c8e171.

TL;DR is – either regenerate your Xcode project using `react-native init` or `react-native upgrade`, or manually add "Run Script Phase" to your Xcode project:

image


image


Note that the location for main.jsbundle has changed and you no longer need to manually run `react-native bundle`, it will be handled by Xcode.

**Comment by shishircc** [comment]
Thanks @frantic ,
I created project again by using react-native init
By default it tries to load the code from developer node server running on my machine. 
If I uncomment the bundle line (//  jsCodeLocation = [[NSBundle mainBundle] URLForResource:@"main" withExtension:@"jsbundle"];) like I used to in earlier releases, the program still asks me to create offline bundle 

How is the automatic offline bundling triggered ?

**Comment by frantic** [comment]
> I created project again by using react-native init

This will be part of 0.14 release, the current version is 0.13.2.

If you want to try this feature before the release, do the following: 

1. `rm -rf node_modules`
2. in your `package.json` change "react-native" version to "v0.14.0-rc"
3. `npm install`
4. `react-native upgrade`
5. Answer Y to all questions

**Comment by chirag04** [comment]
@frantic I generated the project today using the master branch(modified cli). Xcode cannot build the new generated project. i get ``/bin/sh react-native command not found``

However the build works fine when i check ``Run scripts only when installed`` in build phases. Can we enable it by default in the template if it's required?

**Comment by frantic** [comment]
@chirag04 can you open a separate issue for this and assign to me?

**Comment by jaredly** [comment]
Does this mean that the `require("image!some_name")` syntax is no longer supported? It looks like the requires aren't being transformed, and I'm getting 
```
Unable to resolve module image!my_image from /the/path: Invalid directory /the/path/image!my_image
```

**Comment by ajwhite** [comment]
Hmmm also curious about this :point_up: 

Neither [v0.14.0-rc](https://github.com/facebook/react-native/releases/tag/v0.14.0-rc) nor [v0.14.0](https://github.com/facebook/react-native/releases/tag/0.14.0) make mention of this being a breaking change

**Comment by foghina** [comment]
@frantic can you have a look into this? It looks similar to what I'm seeing on travis on the 0.14-stable branch: https://travis-ci.org/facebook/react-native/jobs/89546902#L1272

**Comment by foghina** [comment]
Also cc @martinbigio

**Comment by frantic** [comment]
> Does this mean that the require("image!some_name") syntax is no longer supported? 

@jaredly should just work, seem like a bug. FYI you can still use `{uri: 'some_name'}` (which is actually how `require('image!some_name')` is transformed under the hood).

**Comment by popbee** [comment]
2 questions about the Asset system:  

#1- If I understand correctly, there will be no way to have images updated without updating the native portion of an app?  Can't use mechanisms like **AppHub** to update/add/remove images?

#2- I see the names needs to be statically defined. Would there be a way to have a simple **wildcard pattern**, something like `Anim*.png` or `status_*.png`?

(Not sure this is the right place to ask, please redirect me anywhere more appropriate if not)

**Comment by foghina** [comment]
1. We don't support OTA updates at all yet, but something could be built by the community / services like AppHub.

2. Not sure what you mean. Requiring e.g. `image.png` will return the correct resource for your current configuration (pixel density etc.), e.g. `image@2x.png` or `image@3x.android.png`.  Not sure how wildcards would be useful if this already works?

**Comment by frantic** [comment]
@popbee regarding 1 see #3679 - basically we will load images from the same place as JS, which means you could potentially put your JS bundle and assets into a folder and tell RN to load from it and it should just work.

**Comment by bradbumbalough** [comment]
I had an issue when I tried to offline bundle the app, even after upgrading the `react-native` package with the `react-native upgrade` command, but it wasn't until I manually added the run script task that it is now working. Thanks @frantic.

**Comment by andpor** [comment]
I ran react-native upgrade but obviously couldn't answer Y to all overwrites as majority of existing project config would be lost so I had to be selective. Nonetheless, the run script task was not added.

Separately, the new run script task makes the assumption that the entry point has to be index.ios.js - this is an artificial and limiting constraint in my view . It should be passed in as param to the script. Also, locations it uses are artificial and limiting - they are user project dependent really...

Still, cannot get static images to work with the offline bundle deployment in Xcode...have not even tried Android..this is React 0.18...

**Comment by satya164** [comment]
> I ran react-native upgrade but obviously couldn't answer Y to all overwrites as majority of existing project config would be lost so I had to be selective. Nonetheless, the run script task was not added.

You can press "d" to view the diff of what's changed, and do those changes manually.

> Separately, the new run script task makes the assumption that the entry point has to be index.ios.js - this is an artificial and limiting constraint in my view . It should be passed in as param to the script. Also, locations it uses are artificial and limiting - they are user project dependent really...

The `react-native bundle` command takes the entry file name as an argument.

> Still, cannot get static images to work with the offline bundle deployment in Xcode...have not even tried Android..this is React 0.18...

Try creating a project from scratch and port your changes over. This is the easiest way.

> The Awesome Project in Xcode does not show any references to assets folder and iOS is not finding the static images...I have not idea where the images should be placed and how they should be referenced from Xcode any longer...

You can place them anywhere you want and reference them just like you reference javascript files.

**Comment by andpor** [comment]
@satya164 

1. You can press "d" to view the diff of what's changed, and do those changes manually.

Yes, did that. The run script task was never added to the Xcode project ...

2. The react-native bundle command takes the entry file name as an argument.

Yes. but all documentation scattered around refers to running the pre-canned script react-native-code.sh and supposedly this script is supposed to be added by react-native init or upgrade so as you can image someone will be scratching their head to figure out what is going on unless they go into this sh script and modify things around. Furthermore, since the script does not take any params, the whole script has to be duplicated in order to run it against a different entry point - not a good thing.

3. Try creating a project from scratch and port your changes over. This is the easiest way.

That is what I did with React 0.18 - created Awesome Project from scratch. This in fact created a run script task but I still cannot see how the static images are referenced from Xcode side.

4. You can place them anywhere you want and reference them just like you reference javascript files.

Yes, should be trivial but it does not seem to work that way. In general it works if you load the bundle from the dev server but things do not work if you generate a offline bundle ...images normally resolved via dev server are not picked up in running iOS app.

5. react-native bundle fails if run with --dev false which is what the react-native-code.sh does. If you run it with --dev true - seems to work fine.

screen shot 2016-02-04 at 2 11 32 pm

**Comment by frantic** [comment]
Regarding `react-native upgrade`: unfortunately this process is not streamlined yet, it just relies on yeoman's ability to upgrade generated files. Merging Xcode project files is pain, so if you want to do it manually I have a comment above with a few pictures that describe how to do it.

As for the entry point – we rely on convention over configuration here. If you think that the process of customizing the entry point is common enough, please send a PR. We [have addressed](https://github.com/facebook/react-native/pull/5518) [a similar issue](https://github.com/facebook/react-native/issues/3948) earlier by adding a configuration option.


The bug your experienced regarding whole program optimizations deserves a separate issue and is not related to this PR. I suspect your `node_modules` might be in a bad state, try `rm -rf node_modules && npm install`.
